Allow reading FileInfo from a dummy file instead of the file itself

Myself5 opened the pull request:
On our (see #58) server setup, the mirrorbits server is not serving any files and is only responsible for spreadding the load. Effectively, this means it only needs the files for checksum and size generation, which produces a lot of wasted storage.

we use https://github.com/Myself5/mirrorbits_dummycreator for the dummy creation, its effectively just a json table containing the file information and therefore saves us a lot of storage (5.8MB vs 217GB).

etix commented on scan/scan.go:
```diff
@@ -237,10 +247,43 @@ func (s *sourcescanner) walkSource(conn redis.Conn, path string, f os.FileInfo,
-	d.modTime = f.ModTime()
+
+	if dummyFile {
+		raw, err := ioutil.ReadFile(path)
```
In case someone makes a configuration error, we don't want mirrorbits to load a multi-gigabyte file into memory. Making a custom reader that will do some basic sanity check would be much better. Even better, if the file is not a "dummyFile", then fallback to the standard method. This way, we can have an hybrid repository.

etix commented on scan/scan.go:
```diff
@@ -237,10 +247,43 @@ func (s *sourcescanner) walkSource(conn redis.Conn, path string, f os.FileInfo,
+		if err != nil {
+			fmt.Println(err.Error())
+			fmt.Println("Failed to read file, DELETING!")
+			os.Remove(path)
```
This is way too dangerous and can result in a massive data loss if the redirector is also the main rsync server. Just ignore the file and display an error to the console instead.

etix commented on scan/scan.go:
```diff
@@ -237,10 +247,43 @@ func (s *sourcescanner) walkSource(conn redis.Conn, path string, f os.FileInfo,
+		var c []DummyFile
+		err = json.Unmarshal(raw, &c)
+		if err != nil {
+    		log.Errorf("error decoding json: %v", err)
```
It seems you're mixing space and tabs. Please `go fmt` your code.

etix commented on scan/scan.go:
```diff
@@ -237,10 +247,43 @@ func (s *sourcescanner) walkSource(conn redis.Conn, path string, f os.FileInfo,
+	if dummyFile {
+		raw, err := ioutil.ReadFile(path)
+		if err != nil {
+			fmt.Println(err.Error())
```
Use `log.Error` instead of fmt.

etix commented on scan/scan.go:
```diff
@@ -237,10 +247,43 @@ func (s *sourcescanner) walkSource(conn redis.Conn, path string, f os.FileInfo,
+		d.size =  dfData.Size
+		d.modTime, err = time.Parse("2006-01-02 15:04:05.999999999 -0700 MST", dfData.ModTime)
+		if err != nil {
+			fmt.Println(err)
```
Use `log.Error` instead of fmt.

Myself5 commented:
OK, so I adjusted the fmt calls, and allowed fallback to "normal" mode. For reading I added a basic check if the file is > 1MB (which no JSON file should ever be considering it contains only 5 values) and I ran it through gofmt -w scan/scan.go. For a custom reader I lack the knowledge (thats my first GO project ever), however, I wouldn't see what else we can check for.

Only thing that we might want to consider is removing the log in line 259, as on a hybrid repo that would be "intended" I'd guess.

Havn't run it in production yet either, will see if I get some time to do that on the weekend.

etix commented on scan/scan.go:
```diff
@@ -237,11 +247,51 @@ func (s *sourcescanner) walkSource(conn redis.Conn, path string, f os.FileInfo,
 
+	if dummyFile {
+		fi, _ := os.Stat(d.path)
+		if fi.Size() > 1048576 {
```
Magic numbers appearing in the middle of the code are usually bad. One way to improve this specific piece of code is to add to the `util` package (or even `util` subpackage) something like:
```go
type ByteSize int64

const (
	_           = iota
	KB ByteSize = 1 << (10 * iota)
	MB
	GB
	TB
	PB
)
```
Therefore your check will end up looking like this instead: `fi.Size() > 1 * utils.MB`.

That being said, do we really want to enforce / hardcode a limit? What if someday we'll have much bigger (and valid) files to parse? One way I see is to use a [json.Decoder](https://golang.org/pkg/encoding/json/#NewDecoder) instead. It takes an `io.Reader` and therefore will only read the necessary portion of files (in most cases, way less than 1MB). It will also have the advantage of being faster.

etix commented on scan/scan.go:
```diff
@@ -237,11 +247,51 @@ func (s *sourcescanner) walkSource(conn redis.Conn, path string, f os.FileInfo,
+	if dummyFile {
+		fi, _ := os.Stat(d.path)
+		if fi.Size() > 1048576 {
+			log.Errorf("File is bigger than 1MB, falling back to normal")
```
Is it really an error in case of hybrid dummy file support?

etix commented on scan/scan.go:
```diff
@@ -237,11 +247,51 @@ func (s *sourcescanner) walkSource(conn redis.Conn, path string, f os.FileInfo,
+		raw, err := ioutil.ReadFile(path)
+		if err != nil {
+			log.Errorf(err.Error())
+			log.Errorf("Failed to read file, falling back to normal")
```
`Errorf` uses a Printf-like syntax, therefore your can replace this by:
```go
log.Errorf("Failed to read file: %s", err.Error())
```
But again, do we want to display an error in case of an hybrid repository? What about `log.Debugf`?

etix commented:
> Havn't run it in production yet either, will see if I get some time to do that on the weekend.

I have yet no way to test it in production, therefore if you can do it, it would definitely speedup the merge.

Thanks again for your work :)

Myself5 commented:
Allright, every requirement should be fullfilled now. I made sure that hybrid is working. The only close edge case I could image is, that someone uploads a json file with the exact same struct we use (lets say this: https://paste.myself5.de/d0ULubrZJL.json) in THAT case, the json load process ignores the second entry and just loads the values from the first entry. We might want to consider adding a fallback and moving to "normal" mode if a file like that gets detected, up to you.

etix commented:
Hello,

I don't understand the following statement:

> The only close edge case I could image is, that someone uploads a json file with the exact same struct we use (lets say this: https://paste.myself5.de/d0ULubrZJL.json) in THAT case, the json load process ignores the second entry and just loads the values from the first entry.

Myself5 commented:
Assuming we upload a file like the one I linked, we currently threat it as a dummy file, and read the information we need from the first entry in that json array. We should decide if we want to keep that behaviour, or if we want to treat the file as a "normal" file in case of more than one existing json entry.

etix commented:
Yes, indeed. Another option would be to name these dummy files with a specific extension. For instance, for a file named `vlc-2.2.6-win32.exe` that we want to distribute, mirrorbits can check if a file `vlc-2.2.6-win32.exe.mdata` exists. This file would of course contains the appropriate json. In addition doing hybrid repositories can be easier to manage since you can easily differentiate between complete files and metadata-only files.

Opinions?

darix commented:
why not store the metadata for each file in the DB?
